Code/GraphMol/RGroupDecomposition/RGroupDecomp.cpp

```diff
@@ -0,0 +1,1011 @@
+    labels.insert(label);
+    maxLabel = label + 1;
+    return true;
+  }
+  return false;
+}
+
+bool hasDummy(const RWMol &core) {
+  for (RWMol::ConstAtomIterator atIt = core.beginAtoms(); atIt != core.endAtoms();
+       atIt++) {
+    if((*atIt)->getAtomicNum() == 0) return true;
+  }
+  return false;
+}
+}
+
+bool RGroupDecompositionParameters::prepareCore(RWMol &core,
+                                                const RWMol *alignCore) {
+  const bool relabel = labels & RelabelDuplicateLabels;
+  if (alignCore && (alignment & MCS)) {
+    std::vector<ROMOL_SPTR> mols;
+    mols.push_back( ROMOL_SPTR(new ROMol(core)) );
+    mols.push_back( ROMOL_SPTR(new ROMol(*alignCore)) );
+    MCSResult res = findMCS(mols);
+    if(res.isCompleted()) {
+      const ROMol*m = (const ROMol*)SmartsToMol(res.SmartsString);
+      if(m) {
+        MatchVectType match1;
+        MatchVectType match2;
+        
+        RWMol* m = SmartsToMol(res.SmartsString.c_str());
+        bool target_matched1 = SubstructMatch(core, *m, match1);
+        bool target_matched2 = SubstructMatch(*alignCore, *m, match2);
+        CHECK_INVARIANT(match1.size() == match2.size(),
+                        "Matches should be the same size in prepareCore");
+        
+        if(target_matched1 && target_matched2) {
+          for(size_t i=0;i<match1.size();++i) {
+            int queryAtomIdx1 = match1[i].first;
+            int coreAtomIdx = match1[i].second;
+            int queryAtomIdx2 = match2[i].first;
+            int alignCoreAtomIdx = match2[i].second;
+            CHECK_INVARIANT(queryAtomIdx1==queryAtomIdx2,
+                            "query atoms aren't the same");
+            const Atom *coreAtm = core.getAtomWithIdx(coreAtomIdx);
+            const Atom *alignCoreAtm = alignCore->getAtomWithIdx(alignCoreAtomIdx);
+            int rlabel = alignCoreAtm->getProp<int>(RLABEL);
+            coreAtm->setProp(RLABEL, rlabel);
+          }
+        }
+        delete m;
+      }
+    }
+  }
+  
+  std::set<int> foundLabels;
+  int maxLabel = 0;
+  int nextOffset = 0;
+  for (RWMol::AtomIterator atIt = core.beginAtoms(); atIt != core.endAtoms();
+       atIt++) {
+    Atom *atom = *atIt;
+    if (atom->hasProp(RLABEL)) continue;
+    
+    if (labels & IsotopeLabels) {
+      if (setLabel(atom, rdcast<int>(atom->getIsotope()), foundLabels, maxLabel,
+                   relabel, "IsotopeLabels"))
+        continue;
+    }
+
+    if (labels & AtomMapLabels) {
+      if (setLabel(atom, rdcast<int>(atom->getAtomMapNum()), foundLabels,
+                   maxLabel, relabel, "AtomMapLabels"))
+        continue;
+    }
+
+    if (labels & AtomIndexLabels) {
+      if (setLabel(atom, indexOffset - atom->getIdx(), foundLabels, maxLabel,
+                   relabel, "IndexLabels"))
+        nextOffset++;
+        continue;
+    }
+  }
+
+  indexOffset -= nextOffset;
+
+  std::map<int, int> labels;
+  for (RWMol::AtomIterator atIt = core.beginAtoms(); atIt != core.endAtoms();
```

Comment:
Yes.  If I get rid of the continue's and add an "found" boolean it could fall through and set the labels.